packages/frames/src/Frame.ts

```diff
@@ -0,0 +1,186 @@
+import { MetaTag } from 'next-seo/lib/types';
+
+export enum FrameAspectRatio {
+  SQUARE = '1:1',
+  WIDE = '1.91:1',
+}
+
+export enum FrameButtonAction {
+  /**
+   * The button will trigger a post request to the postUrl.
+   * By default, the postUrl is used as the target URL if target is not provided.
+   */
+  POST = 'post',
+  /**
+   * The button will redirect to the target URL.
+   */
+  LINK = 'link',
+}
+
+export interface FrameButton {
+  /**
+   * The label to be displayed in the button.
+   */
+  label: string;
+  /**
+   * - post: The button will trigger a post request to the postUrl.
+   * - link: The button will redirect to the target URL.
+   * @default post
+   */
+  action?: FrameButtonAction;
+  /**
+   * A target URL for the button. It uses the postUrl if not provided.
+   */
+  target?: string;
+}
+
+interface FrameConstructorProps {
+  /**
+   * The title to be displayed in the frame. Required.
+   */
+  title: string;
+  /**
+   * The URL of the image to be displayed in the frame. It should be a URL. Required.
+   */
+  image: string;
+  /**
+   * The URL to be used in the post request. Required.
+   */
+  postUrl: string;
+  /**
+   * Aspect-ratio of the image. Required.
+   */
+  aspectRatio: FrameAspectRatio;
+  /**
+   * An ordered list of buttons to be displayed in the frame.
+   */
+  buttons?: FrameButton[];
+}
+
+/**
+ * Represents a Farcaster frame.
+ * Documentation: https://docs.farcaster.xyz/learn/what-is-farcaster/frames
+ * Specification: https://docs.farcaster.xyz/reference/frames/spec
+ */
+export class Frame {
+  /**
+   * The title to be displayed in the frame. Required.
+   */
+  title: string;
+  /**
+   * The URL of the image to be displayed in the frame. It should be a URL. Required.
+   */
+  image: string;
+  /**
+   * The URL to be used in the post request. Required.
+   */
+  postUrl: string;
+  /**
+   * Aspect-ratio of the image. Required.
+   */
+  aspectRatio: FrameAspectRatio;
+  /**
+   * An ordered list of buttons to be displayed in the frame.
+   */
+  buttons?: FrameButton[];
+
+  constructor(props: FrameConstructorProps) {
+    this.title = props.title;
+    this.image = props.image;
+    this.postUrl = props.postUrl;
+    this.aspectRatio = props.aspectRatio;
+    this.buttons = props.buttons;
+  }
+
+  /**
+   * Renders the frame as complete HTML markup.
+   *
+   * @example
+   * export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+   *   const frame = new Frame({ ... });
+   *   return res.status(200).send(frame.render());
+   * }
+   *
+   * @returns The HTML representation of the frame.
+   */
+  public render(): string {
+    const metaTags = this.toMetaTags().map(Frame.renderTag).join('');
+    return `<html><head><title>${this.title}</title>${metaTags}</head><body></body></html>`;
```

Comment:
No DTD? I think adding HTML5 DTD will be good a choice.

What does it mean that the body is empty?

> Should return a valid HTML <body>, in case the user clicks through to the frame in a browser.

Should this have an `<img />` tag with the image to display in case that is not show as a frame?

Maybe not. I am asking out of curiosity.

Reply:
Having the `<img />` wouldn't hurt.

I'm not sure what Farcaster means with "cliking through the frame", though. Maybe just typing out the endpoint URL and accessing the page directly.

Clicking _on_ a frame will direct the user to the linked page, just as a standard link. Just by interacting with a frame in a standard way, you can't be sent to the frame code, really.

With that said, adding DTD and image tag does seem like a good idea to me.